### ISSUE TITLE
For a type like "CompletionTriggerKind = 1 | 2 | 3", can servers assume they will never get another value?

### ISSUE THREAD

**Issue by DanTup** [microsoft/language-server-protocol#686]
There are a lot of types in the spec that are made up of constants, for example:

```ts
CompletionTriggerKind = 1 | 2 | 3
```

The spec says that this type will be passed when `contextSupport` is set to `true`. It's not clear whether in the future some clients may support an additional value (`4`) or not, and whether servers should be written to accommodate that.

This could mean the different between modelling this type as a number on the server (can be anything) or an enum that would only allow the specified values (in which case, an unknown value may result in a failure to deserialise the message).

Assuming it's possible there may be a `4` in future, I think the `CompletionTriggerKind` should be annotated in some way that it's value might *not* be `1 | 2 | 3`.

**Comment by DanTup** [comment]
These are the items I can see in the spec where the type *looks* to be fixed to the specified list:

- CompletionTriggerKind
- FailureHandlingKind
- InsertTextFormat
- MarkupKind
- ResourceOperationKind

SymbolKind also has a restricted type, however it's explicitly handled by the client advertising which it supports (and in doing so, guarantees it will handle unknown values).

The other types all tend to use `string` or `number` in the classes that hold them, which indicates they accept any value. For now, I'm assuming it's not valid for a client (or server) to send values for the types listed above that are not listed in the spec (and that they will not be extended in future). I'm not sure that's intended though (for ex. SumbolKind seems like the sort of thing that

**Comment by dbaeumer** [comment]
Your assumption is correct and not. No values should be passed around that are not in the spec unless explicitly stated. However these enums can have additional values in the future however this needs to be guarded by a capability to ensure that the other sides support them. An example is the `SymbolKind` which we extended. However a client needs to signal that to the server.

OK to close?

**Comment by DanTup** [comment]
> However these enums can have additional values in the future however this needs to be guarded by a capability to ensure that the other sides support them

My understanding is that was why some types were typed as `num`/`string` instead of the enum type. But now I'm confused. If I copied the definition of `CompletionTriggerKind` from the spec, and wrote this code:

```ts
type CompletionTriggerKind = 1 | 2 | 3;

export interface CompletionContext {
	triggerKind: CompletionTriggerKind;
}

let a: CompletionContext = null;

if (a.triggerKind === 1) {
	print("1");
} else if (a.triggerKind === 2) {
	print("2");
} else if (a.triggerKind === 3) {
	print("3");
} else {
	print(a.triggerKind); // a.triggerKind is never
}
```

The type of `a.triggerKind` at the end here is `never`, which makes it seem like this can't possibly happen. It seems like the definition of `CompletionTriggerKind` here is incorrect if it may be extended in future. If there is a chance that a client could send a `4` it seems dangerous to use this definition.

In that case, I think it would be better to remove the `CompletionTriggerKind` and type it as `number` in the `CompletionContext.triggerKind` property?

**Comment by dbaeumer** [comment]
You are talking about coding against the API. I still see no problem typing it as `1 | 2 | 3` since it makes things more clear. Assuming we add another kind `4`. If your client code doesn't set a capability that `4` is supported a server should not send for, so your code is still correct. If you opt in supporting `4` then of course you need to change the if / else cascade as well.

**Comment by DanTup** [comment]
@dbaeumer I mean that the code above is probably not what server authors will write, because their type system may say that the `else` case is not possible. Writing it as `1 | 2 | 3` makes it really easy to write code that never handles values outside of that (since the type system says they're not possible) and therefore end up with servers that aren't very forwards-compatible.

In normal TypeScript code (ignoring LSP), probably you wouldn't type something as `1 | 2 | 3` if you knew that in future the code might get a 4 (before you've been able to deploy updated code).

It also makes it difficult when generating code to know which cases are really closed sets (eg. other values can immediately reject with INVALID_PARAMS during deserialisation) and which are not (though now I'm questioning whether anything at all is actually a closed set).

**Comment by dbaeumer** [comment]
I see the point however if outside values are handled nicely by the server code like this can be written, which makes this even more clear IMO

```ts
if (....) {
} else if (...) {
} else {
   const outside = a.triggerKind;
   pint(outside);
}
```

**Comment by DanTup** [comment]
Sure, but my concern was that it was easily to *accidentally* not handle it because the types say that it can't happen. If you write the code above and hover your mouse over `triggerKind` you see this:

<img width="422" alt="Screenshot 2019-03-27 at 4 48 20 pm" src="https://user-images.githubusercontent.com/1078012/55095494-27943000-50b0-11e9-92a4-d2b041663d96.png">

It would be easy to assume this cannot happen and delete the else (the type says `never` after all). It doesn't seem crazy that you may convert the spec into another language which tells you this is unreachable code outright (I don't know if any do this, but it seems valid).

If another property was changed from `string | number` to `string | number | boolean` it would be considered a breaking change and would (presumably) be guarded by the client/server having to advertise that it supports the new `boolean` option in capabilities before they are sent. It seems like this is the same thing.

Eg. when `3` was added to `CompletionTriggerKind` there was nothing done to protect against servers that might have done this (which seems completely valid given that the spec specifically says the value is `1 | 2` and *not* `number`):

```ts
if (triggerKind == CompletionTriggerKind.Invoked) {
// Handle invoked
} else {
// Handle TriggerCharacter
}
```

and then clients send them a 3 and it behaves as a 2.

Anyway, I think you understand what I'm saying so if you think it's fine as-is, I won't keep banging on :) I'll change my code-gen not to assume unions of literal strings/numbers are that fixed set. I guess in theory every one of them could be extended in future.

**Comment by Aerijo** [comment]
It sounds like there will be capability flags guarding any extensions to value sets, so servers written with a fixed value set are fine to do so because they won't be opting into the capability with the extended values. Is the clarification tag for adding this explicitly to the spec?

**Comment by dbaeumer** [comment]
The wat we currently go is not a flag be value but that the client announces its used values if we extend them

**Comment by Aerijo** [comment]
OK, not necessarily flags, but in

> this needs to be guarded by a capability to ensure that the other sides support them.

Are you saying that extra values will not be used unless both sides indicate support for them? Which means that a server/client can just treat the set as fixed until it decides to support the extended set.

**Comment by dbaeumer** [comment]
The client announces which value set it supports. The server is supposed to not use values outside the client set. This being said a client should protect itself against unknown properties if there is a meaningful default (for example I default an severity to error if its value is unknown in the VS Code client).

**Comment by dbaeumer** [comment]
The reason why this behaves the way it currently does is that LSP is versioned using these capability flags and not version numbers. This is something we decided a long time ago and is nothing we can change now (and I still think it was the right approach since it allows clients and server gradually adopt new features).

I will close the issue since I don't see a way of changing this now. I also made an effort for the 3.16 spec to ensure the right property announcement in the initialize call.
